applications/vanilla/models/class.categorymodel.php

```diff
@@ -2801,4 +2805,126 @@ public function recalculateTree() {
+                    ->where('CategoryID', $targetID)
+                    ->put();
+            }
+        }
+
+        // Update the cache.
+        $categoriesToUpdate = self::instance()->getWhere(['CategoryID' => $updatedCategories]);
+        foreach ($categoriesToUpdate as $current) {
+            $currentID = val('CategoryID', $current);
+            $countAllDiscussions = val('CountAllDiscussions', $current);
+            $countAllComments = val('CountAllComments', $current);
+            self::setCache(
+                $currentID,
+                ['CountAllDiscussions' => $countAllDiscussions, 'CountAllComments' => $countAllComments]
+            );
+        }
+    }
+
+    /**
+     * Move upward through the category tree, incrementing aggregate post counts.
+     *
+     * @param int|string|object|array $category
+     */
+    public static function incrementAggregateCounts($category) {
+        self::adjustAggregateCounts($category, 1);
+    }
+
+    /**
+     * Move upward through the category tree, decrementing aggregate post counts.
+     *
+     * @param int|string|object|array $category
+     */
+    public static function decrementAggregateCounts($category) {
+        self::adjustAggregateCounts($category, -1);
+    }
+
+    /**
+     * Recalculate all aggregate post count columns for all categories.
+     *
+     * @return void
+     */
+    private static function recalculateAggregateCounts() {
+        // First grab the max depth so you know where to loop.
+        $depth = Gdn::sql()
+            ->select('Depth', 'max')
+            ->from('Category')
+            ->get()
+            ->firstRow(DATASET_TYPE_ARRAY);
+        $depth = (int)val('Depth', $depth, 0);
+
+        if ($depth === 0) {
+            return;
+        }
+
+        $prefix = Gdn::database()->DatabasePrefix;
+
+        // Initialize with self count.
+        Gdn::sql()
+            ->update('Category')
+            ->set('CountAllDiscussions', 'CountDiscussions', false)
+            ->set('CountAllComments', 'CountComments', false)
+            ->put();
+
+        while ($depth > 0) {
+            $sql = "update {$prefix}Category c
+                    join (
+                        select
+                            c2.ParentCategoryID,
+                            sum(CountAllDiscussions) as CountAllDiscussions,
+                            sum(CountAllComments) as CountAllComments
+                        from {$prefix}Category c2
+                        where c2.Depth = :Depth
+                        group by c2.ParentCategoryID
+                    ) c2 on c.CategoryID = c2.ParentCategoryID
+                set
+                    c.CountAllDiscussions = c.CountAllDiscussions + c2.CountAllDiscussions,
+                    c.CountAllComments = c.CountAllComments + c2.CountAllComments
+                where c.Depth = :ParentDepth";
+            Gdn::database()->query($sql, [':Depth' => $depth, ':ParentDepth' => ($depth - 1)]);
```

Comment:
Implementing the method recommended here: https://github.com/vanilla/vanilla/issues/3902#issuecomment-279808127